docs/rfcs/0003-sql-datastore.md

```diff
@@ -0,0 +1,280 @@
+- able to keep the advantage of schemaless pattern
+
+## Ability to index a specific attribute of the JSON field
+
+We have 2 points which have to be considered.
+
+1. UUID as primary keys should not affect queries performance
+
+While MySQL does not have UUID data type, since we're using `application side UUID generate` pattern, (which mean to MySQL, those ids are true random UUID) store those ids under `VARCHAR(32)` data type is costly for both read and write operation due to the indexes does not work, using `UUID_TO_BIN` & `BIN_TO_UUID` with `swap_flag` and store data under `BINARY(16)` would help.
+
+From MySQL docs
+```
+  o If swap_flag is 0, the two-argument form is equivalent to the
+    one-argument form. The binary result is in the same order as the
+    string argument.
+
+  o If swap_flag is 1, the format of the return value differs: The
+    time-low and time-high parts (the first and third groups of
+    hexadecimal digits, respectively) are swapped. This moves the more
+    rapidly varying part to the right and can improve indexing
+    efficiency if the result is stored in an indexed column.
+
+```
+
+PostgreSQL has UUID type to store this kind of data and support it as primary key as well.
+
+MySQL: 👍 PostgreSQL: 👍
+
+2. Create indexes for JSON attributes without affecting the schemaless advantage
+
+For MySQL, we could use `CREATE INDEX` to create secondary indexes on generated columns which stored JSON attributes as an indirect way to indexes JSON attributes.
+In order to keep the advantage of the schemaless pattern, we will use `virtual generated columns` instead of `stored generated columns` (which will physically store along with other columns of table). __The virtual generated columns wouldn't be generated on READ as long as we keep all generated columns as part of some secondary indexes__, which reduce the cost of recomputing from READ operation (note that computing virtual columns value cost on WRITE remains).
+
+Sample secondary indexes creation commands as follow
+
+```sql
+mysql> CREATE INDEX idx ON applications ( ( CAST( data->>"$.name" AS CHAR(10) ) ) );
+# OR
+mysql> CREATE INDEX idy ON applications ( (JSON_VALUE(data, '$.name' RETURNING CHAR(10))) );
+```
+
+ref:
+- https://dev.mysql.com/doc/refman/8.0/en/create-table-secondary-indexes.html
+- https://dev.mysql.com/doc/refman/8.0/en/create-index.html
+
+For PostgreSQL, we also have `CREATE INDEX` statement for this task too. The good thing is, PostgreSQL treats indexes for JSON attributes as same as indexes for normal columns so that we don't have to worry about generated columns or something else be added to our schema.
+
+Sample indexes creation commands as follow
+
+```sql
+postgres=> CREATE INDEX idx ON test.applications ((data->>'name'));
+```
+
+ref: https://www.postgresql.org/docs/13/datatype-json.html
+
+Since both MySQL and PostgreSQL have multi-columns indexes, it's okay if we want to add index to multi attributes of JSON data.
+
+MySQL: 👍 PostgreSQL: 👍
```

Comment:
Maybe PostgreSQL should have 2 👍  (👍 👍 ) because of its simplicity.

Reply:
then it should be in this `Community supports` section 😄 
https://github.com/pipe-cd/pipe/pull/1612/files#diff-61adf68d03c8a5d48c38d1ccad3f1f95bf0b6948f603525cebb5cb5fdcf78e1bR276